firedrake/slate/slac/kernel_builder.py

```diff
@@ -653,54 +745,247 @@ def slate_call(self, prg, temporaries):
+                 and 0<=i_6<=n}""",
+            ["""{x}[i_0] = -{b}[i_0] {{id=x0}}
+                {A_on_x}[:] = action_A({A}[:,:], {x}[:]) {{dep=x0, id=Aonx}}
+                <> r[i_3] = {A_on_x}[i_3]-{b}[i_3] {{dep=Aonx, id=residual0}}
+                {p}[i_4] = -r[i_4] {{dep=residual0, id=projector0}}
+                <> rk_norm = 0. {{dep=projector0, id=rk_norm0}}
+                rk_norm = rk_norm + r[i_5]*r[i_5] {{dep=projector0, id=rk_norm1}}
+                for i_6
+                    {A_on_p}[:] = action_A_on_p({A}[:,:], {p}[:]) {{dep=Aonp0, id=Aonp, inames=i_6}}
+                    <> p_on_Ap = 0. {{dep=Aonp, id=ponAp0}}
+                    p_on_Ap = p_on_Ap + {p}[i_2]*{A_on_p}[i_2] {{dep=ponAp0, id=ponAp}}
+                    <> projector_is_zero = abs(p_on_Ap) < 1.e-16 {{id={preconverged_criterion_dep}, dep=ponAp}}
+             """.format(**str2name),
+             corner_case,
+             """    <> alpha = rk_norm / p_on_Ap {{dep={preconverged_criterion_id}, id=alpha}}
+                    {x}[i_7] = {x}[i_7] + alpha*{p}[i_7] {{dep=ponAp, id=xk}}
+                    r[i_8] = r[i_8] + alpha*{A_on_p}[i_8] {{dep=xk,id=rk}}
+                    <> rkp1_norm = 0. {{dep=rk, id=rkp1_norm0}}
+                    rkp1_norm = rkp1_norm + r[i_9]*r[i_9] {{dep=rkp1_norm0, id={stop_criterion_dep}}}
+             """.format(**str2name),
+             stop_criterion,
+             """    <> beta = rkp1_norm / rk_norm {{dep={stop_criterion_id}, id=beta}}
+                    rk_norm = rkp1_norm {{dep=beta, id=rk_normk}}
+                    {p}[i_10] = beta * {p}[i_10] - r[i_10] {{dep=rk_normk, id=projectork}}
+                    {A_on_p}[i_12] = 0. {{dep=projectork, id=Aonp0, inames=i_6}}
+                end
+                {output}[i_11] = {x}[i_11] {{dep=Aonp0, id=out}}
+             """.format(**str2name)],
+            [*args,
+             loopy.TemporaryVariable(str2name["x"], dtype, shape=shape, address_space=loopy.AddressSpace.LOCAL, target=loopy.CTarget()),
+             loopy.TemporaryVariable(A_on_x_name, dtype, shape=shape, address_space=loopy.AddressSpace.LOCAL),
+             loopy.TemporaryVariable(A_on_p_name, dtype, shape=shape, address_space=loopy.AddressSpace.LOCAL),
+             loopy.TemporaryVariable(str2name["p"], dtype, shape=shape, address_space=loopy.AddressSpace.LOCAL)],
+            target=loopy.CTarget(),
+            name=name,
+            lang_version=(2018, 2),
+            silenced_warnings=["single_writer_after_creation", "unused_inames"])
+
+        # set the length of the indices to the size of the temporaries
+        knl = loopy.fix_parameters(knl, n=shape[0])
+
+        # update gem to pym mapping
+        # by linking the actions of the matrix-free solve kernel
+        # to the their pymbolic variables
+        _ = ctx.pymbolic_variable(expr.Aonx, knl.callables_table[name].subkernel.id_to_insn["Aonx"].assignees[0].subscript.aggregate.name)
```

Comment:
Why the `_`? If you want to discard the return value can't you just delete `_ = `?

Reply:
After the cleanup of the TSFC branch, I have now changed this bit anyways. But yes you were right.